### ISSUE TITLE
Implement a concept of "Required" properties

### ISSUE THREAD

**Issue by ryanbrandenburg** [dotnet/corefx#38492]
### Current situation
There's no way to indicate that if an object doesn't have a particular property that serialization should fail.

For example:
```c#
public class Person
{
    [Required]
    public string Name { get; set; }
    public int Age { get; set; }
}
```
```json
{
    "Age": 46
}
```

Trying to deserialize the above json as a `Person` will succeed, even though it's been indicated that the name attribute is required, and is missing.

### Describe the solution you'd like
System.Text.Json should respect either `System.ComponentModel.DataAnnotations.RequiredAttribute` or a new attribute (likely extending `System.Text.Json.JsonAttribute`) and error when an attempt is made to serialize an object which doesn't have that property.

### Describe alternatives you've considered
Using `System.ComponentModel.DataAnnotations.RequiredAttribute` might be troublesome because it could indicate Requirement in a different context. The alternative is that users implement their own "Required"  functionality (and likely get it wrong or incomplete).

CC @pranavkm who I talked about this with.

**Comment by Clockwork-Muse** [comment]
.... if `Name` is a non-nullable reference type, wouldn't that indicate that it was required?

So:
```csharp
public class Person
{
    public string Name { get; set; }
    public int Age { get; set; }
}
```
would fail the deserialize, but:
```csharp
public class Person
{
    public string? Name { get; set; }
    public int Age { get; set; }
}
```
... would let it succeed.

**Comment by scalablecory** [comment]
If we implement `Required` we'd also want to implement all the other validation attributes. I'm not sure that System.Text.Json is the correct place to do this.

**Comment by pranavkm** [comment]
Having another look at https://github.com/dotnet/corefx/issues/37485, it looks like there's already another issue that is effectively asking for the same thing. Perhaps we should just close this and use the other issue to track work here?

**Comment by ahsonkhan** [comment]
I believe we should have the support for `Required` (similar to how we have the `JsonIgnore` attribute), but this will likely be something we would address in the future.

This issue talks more directly to the issue at hand, so I would prefer keeping this one as a feature request.

cc @steveharter

**Comment by ahsonkhan** [comment]
From @pranavkm (https://github.com/dotnet/corefx/issues/37485#issuecomment-501482970):

> @JamesNK, did you mean [JsonPropertyRequired](https://www.newtonsoft.com/json/help/html/JsonPropertyRequired.htm)

> MVC users have often asked for a way to identify when a property isn't on the wire. While @steveharter's solutions are acceptable for the trivial cases, it really doesn't scale well if you want the behavior to apply in general. Plus, you're forced to manually perform the checks which seems cumbersome. 

> Could we consider introducing an attribute like Json.NET to enforce this behavior?

**Comment by steveharter** [comment]
Required was discussed but didn't meet the bar for 3.0 in lieu of other features. A workaround is to create a custom converter (pending feature) and override the Read method.

**Comment by pranavkm** [comment]
> Required was discussed but didn't meet the bar for 3.0 in lieu of other features 

Fair enough. I think we're ok with that for this release. We can keep this issue around to track the feature work for the Future.

**Comment by Tragetaschen** [comment]
Virtually all my model's properties have Newtonsoft Json's  `[…, JsonProperty(Required = Required.Always)]`, which that input formatter honors. This is blocking my System.Text.Json adoption.
